posthog/models/tagged_item.py

```diff
@@ -0,0 +1,50 @@
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+
+from posthog.models.utils import UUIDModel
+
+
+class EnterpriseTaggedItem(UUIDModel):
+    """
+    Taggable describes global tag-object relationships.
+
+    Note: This is an EE only feature, however the model exists in posthog so that it is backwards accessible from all
+    models. Whether we should be able to interact with this table is determined in the `TaggedItemSerializer` which
+    imports `EnterpriseTaggedItemSerializer` if the feature is available.
+
+    Today, tags exist at the model-level making it impossible to aggregate, filter, and query objects appwide by tags.
+    We want to deprecate model-specific tags and refactor tag relationships into a separate table that keeps track of
+    tag-object relationships.
+
+    Models that had in-line tags before this table was created:
+    - ee/models/ee_event_definition.py
+    - ee/models/ee_property_definition.py
+    - models/dashboard.py
+    - models/insight.py
+
+    Models that are taggable throughout the app:
+    - models/dashboard.py
+    - models/insight.py
+    - models/event_definition.py
+    - models/property_definition.py
+    - models/action.py
+
+    https://docs.djangoproject.com/en/4.0/ref/contrib/contenttypes/#generic-relations
+    """
+
+    tag: models.SlugField = models.SlugField()
+    team: models.ForeignKey = models.ForeignKey("Team", on_delete=models.CASCADE)
+    color: models.CharField = models.CharField(max_length=400, null=True, blank=True)
+
+    content_type: models.ForeignKey = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    # Primary key value of related model. Query by this to get all tags for specific model. This is a charfield because
+    # there we don't have a standard way of storing objects. Some models use positive integer ids and others use UUID's.
+    object_id: models.CharField = models.CharField(max_length=400)
+    content_object: GenericForeignKey = GenericForeignKey("content_type", "object_id")
```

Comment:
I'm not sure `GenericForeignKey` is the right approach here. I feel like polymorphism can be a pretty big pain. For example, down the road we might want a query that gets the name of everything tagged by X, but an insight might have a "name" column called `title` and dashboard might have a "name" column called `name`. Having it setup with a single `object_id` column can make this kind of painful.

Because there are only so many "taggable" things in PostHog, I wonder if it makes sense to just have columns like `dashboard_id`, `insight_id` etc and some validation to ensure only a single one is set.

I could be convinced otherwise, but that's my gut. If we do decide to move forward with `GenericForeignKey`, I think we should look at adding an object type enum, so we can keep track of which type of object the `GenericForeignKey` is pointed at.